src/librustc_trans/trans/debuginfo.rs

```diff
@@ -1156,12 +1156,42 @@ pub fn get_cleanup_debug_loc_for_ast_node<'a, 'tcx>(cx: &CrateContext<'a, 'tcx>,
         }
     }
 
-    NodeInfo {
+    NodeIdAndSpan {
         id: node_id,
         span: cleanup_span
     }
 }
 
+#[derive(Copy, Clone, PartialEq, Eq)]
+pub enum DebugLocation {
+    DebugLoc(ast::NodeId, Span),
+    NoDebugLoc
+}
+
+impl DebugLocation {
+    pub fn from_expr(expr: &ast::Expr) -> DebugLocation {
+        DebugLocation::DebugLoc(expr.id, expr.span)
+    }
+
+    pub fn from_opt(data: Option<NodeIdAndSpan>) -> DebugLocation {
+        match data {
+            Some(NodeIdAndSpan { id, span }) => DebugLocation::DebugLoc(id, span),
+            None => DebugLocation::NoDebugLoc
+        }
+    }
```

Comment:
Perhaps have a ToDebugLocation trait for these fns and `NodeIdAndSpan::to_source_location`